Corrections to Latency Histogram Calculations based on DBA_HIST_EVENT_HISTOGRAM

davidkurtz opened the pull request:
1. Corrections to Latency Histogram Calculations based on DBA_HIST_EVENT_HISTOGRAM
The duration of the histogram bucket in some summary reports (was being calculated as half the difference in the upper limit for the current and previous lower bucket.  Therefore, these reports were returning incorrect values.  
Also, not all bucket sizes will be present for all events in all snapshots so we cannot use the LAG() function approach either.  Therefore it is safer to use 3/4 of the bucket size (because the upper limit of the next lower bucket will be 1/2 size of this bucket).
In some places, the code was written to assume the smallest bucket was 1ms, but in 19c has smaller buckets (down to .000976563ms).  The code now looks up the smallest known bucket size in DBA_HIST_EVENT_HISTORY and uses 3/4 of that value for larger buckets.

WAIT_CLASS_ID has been removed from the partition by clauses in the analytic functions because they already contain EVENT_ID.  

Oracle reports a histogram with a wait_timi_milli of 1.8014E+16.  This is 570,000 years!  So, it is effectively infinite.  The excessive value for the bucket size distorts the calculations so instead, it is capped at 2 the value of the next bucket.  This continues the geometric progression that every bucket is twice as long as the previous, and still counts the very long waits that fall into the highest bucket.

2. Also: Latch number and address has been added to the latch report in 2a.  Although it is not referenced elsewhere in EDB360, it can be extracted from p1 and p2 in the ASH data that is extracted to a flat file, so it would be useful to be able to look it up somewhere in EDB360.

davidkurtz committed:
Add latch number and name to report in 2a

latch number and address can be extracted from p1 and p2 in the ASH data that is extracted to a flat file, so it would be useful to be able to look it up somewhere in EDB360.

davidkurtz committed:
Correct calculation of average wait time

The average wait time calculated from DBA_HIST_EVENT_HISTOGRAM should be half the sum (not the difference) of the current and previous wait time bucket size.  Also added NVL() to deal with lag from first bucket being null.

davidkurtz committed:
Corrections to Latency Histogram Calculations

Calculation of average_wait_time_milli changed to calculate duration as average of bucket upper limit and upper limit of previous bucket, or half of bucket upper limit if none.
All charts and summaries now use essentially the same formulae
CASE statements in 4i where wait_timi_milli = 1 no longer correct because smaller wait times now reported by DBA_HIST_EVENT_HISTOGRAM.

davidkurtz committed:
Corrections to average wait time for event histogram buckets

The duration of the histogram bucket in some summary reports (was being calculated as half the difference in the upper limit for the current and previous lower bucket.  Therefore, these reports are returning incorrect values.  The average should be calculated as the average of the two, so it should be half the sum.  Where there isn't a previous lower bucket it should be half the upper limit on the bucket.  The lag function returns null, so NVL(...,0) has been added.
WAIT_CLASS_ID has been removed from the partition by clauses in the analytic functions because they already contain EVENT_ID.

davidkurtz committed:
Corrections to Latency Histogram Calculations based on DBA_HIST_EVENT_HISTOGRAM

Oracle reports a histogram with a wait_timi_milli of 1.8014E+16.  This is 570,000 years!  So, it is effectively infinite.  The excessive value for the bucket size distorts the calculations so I have just excluded it from the queries by arbitarily adding ...AND wait_time_milli < 1e9

abelmacias commented:
I do not like the idea to exclude .AND wait_time_milli < 1e9 as it probably tells when an event caused a hang.
I would rather cap it that anything above 4sec to take is a 8sec flat.

The idea of decode latch is to decode the "latch free" waitevent when it happens in col 4 and 5. 
Could be to add to "latch free" the name of the latch with the longest elapsed time during that period.

davidkurtz commented:
Latch analysis is something I am going to look at, and yes profile it something like what happens in cols 4-6.  For now, I am just adding extra information from v$latch to the existing report in 2a.

davidkurtz committed:
Change to use all event histogram buckets but cap the max value

Adjustment to wait time calculation.  It now uses all the buckets, but caps the bucket time for the highest bucket.  All the other buckets are twice the duration of the preceding bucket, the maximum bucket is 1.8E16ms, effectively infinte.  So this value is capped at 1.5x the bucket below so that it follows the progression (it works out at 6291456).
It probably won't make any difference to any of the reports because only Idle waits seem to record waits in this bucket.

davidkurtz committed:
Correct calculation of cap of infinite histogram bucket

The maximum event histogram bucket is 1.8e16 ms.  Instead it will be capped at twice the next largest bucket (2 * 4194304ms), thus following the progression.  The notional midpoint of this bucket will then still be calculated later as 3/4 of this value.

mauropagano commented:
This is great, thanks David!
